src/puppetlabs/puppetdb/cli/benchmark.clj

```diff
@@ -472,27 +431,156 @@
+                                       (println-err (trs "Waiting up to 10s for shutdown"))
+                                       ;; Do we need to make this longer b/c CI mess?
+                                       (.join fill-queue 10000)
+                                       (not (.isAlive fill-queue)))
+                                     (do
+                                       (println-err
+                                        (trs "Unable to verify queue thread has stopped"))
+                                       false)))]
+                  (delete-dir-or-report tmpdir)
+                  result))
+          stopper #(deref stop)]
+      (.addShutdownHook (Runtime/getRuntime) (Thread. stopper))
+      (.start fill-queue)
+      stopper)
+    (catch Exception ex
+      (delete-dir-or-report tmpdir)
+      (throw ex))))
+
+
+(defn cycle-commands
+  "Populates the message queue with messages, and then cycles through
+  those messages, sending and altering each one every time it
+  reappears at the head of the queue.  Returns a stop function of no
+  arguments that can be used to effect an orderly shutdown."
+  [numhosts run-interval rand-perc simulation-threads
+   rate-monitor-ch command-send-ch
+   pdb-host catalogs reports facts]
+  ;; All of the work that needs to be completed on a SIGINT, etc. is
+  ;; coordinated via shutdown hook(s) waiting on daemon thread(s) --
+  ;; because after a signal, only daemon threads, shutdown hooks, and
+  ;; finalizers will still run, and daemon threads will only run until
+  ;; all of the hooks and finalizers are finished.
+  (let [preftmp (get-path (or (System/getenv "TMPDIR")
+                              (System/getProperty "java.io.tmpdir")))
+        tmpdir (Files/createTempDirectory preftmp
+                                          "pdb-bench-"
+                                          (into-array FileAttribute []))
+        mq (try
+             (let [mdir (doto (.resolve tmpdir "q") fs/mkdir)]
+               (chan (message-buffer mdir numhosts)))
+             (catch Exception ex
+               (delete-dir-or-report tmpdir)
+               (throw ex)))
+        stop-fill (start-filling-queue mq numhosts run-interval pdb-host
+                                       catalogs reports facts
+                                       tmpdir)
+        rate-limited-mq-out (chan)
+        channels [rate-limited-mq-out mq rate-monitor-ch command-send-ch]
+        stop (delay
+              ;; Currently assumes that it's ok to leave already
+              ;; in-flight "channel takes" running.  i.e. stop
+              ;; precipitates a stop, but doesn't hang around for
+              ;; anything but the queue deletion.  Also, we don't
+              ;; bother to try to close any of the channels during
+              ;; shutdown shutdown because all those (non-daemon)
+              ;; threads should be dead.
+              (let [close-err? (not-every? identity
```

Comment:
What's the impact of this, seems like nothing to me, since the message is removed from disk before it's dequeued. It either gets dequeued and shutdown before it can submit it, or it gets shutdown after it gets submitted. Either way I think there's no impact right?